homeassistant/components/goodwe/switch.py

```diff
@@ -0,0 +1,131 @@
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN, KEY_DEVICE_INFO, KEY_INVERTER
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class GoodweSwitchEntityDescriptionBase:
+    """Required values when describing Goodwe switch settings."""
+
+    icon_on: str
+    icon_off: str
+
+
+@dataclass(frozen=True)
+class GoodweNumberEntityDescription(
+    SwitchEntityDescription, GoodweSwitchEntityDescriptionBase
+):
+    """Class describing Goodwe switch entities."""
+
+
+SWITCHES = (
+    # Export limit Enabled
+    GoodweNumberEntityDescription(
+        key="grid_export",
+        translation_key="grid_export",
+        icon_on="mdi:transmission-tower-off",
+        icon_off="mdi:transmission-tower-export",
+        entity_category=EntityCategory.CONFIG,
+        device_class=SwitchDeviceClass.SWITCH,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the inverter switch entities from a config entry."""
+    inverter = hass.data[DOMAIN][config_entry.entry_id][KEY_INVERTER]
+    device_info = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE_INFO]
+
+    entities = []
+
+    for description in SWITCHES:
+        try:
+            current_value = await inverter.read_setting(description.key)
+        except (InverterError, ValueError):
+            # Inverter model does not support this setting
+            _LOGGER.debug("Could not read inverter setting %s", description.key)
+            continue
+
+        entities.append(
+            InverterSwitchEntity(device_info, description, inverter, current_value == 1)
+        )
+
+    async_add_entities(entities)
+
+
+class InverterSwitchEntity(SwitchEntity):
+    """Inverter switch setting entity."""
+
+    entity_description: GoodweNumberEntityDescription
+
+    _attr_should_poll = False
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        device_info: DeviceInfo,
+        description: GoodweNumberEntityDescription,
+        inverter: Inverter,
+        current_state: bool,
+    ) -> None:
+        """Initialize the inverter switch setting entity."""
+        self.entity_description = description
+        self._attr_unique_id = f"{DOMAIN}-{description.key}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+        self._attr_is_on = current_state
+        self._inverter: Inverter = inverter
+
+    @property
+    def icon(self):
+        """Return the icon to be displayed for the switch."""
+        if self._attr_is_on:
+            return self.entity_description.icon_on
+
+        return self.entity_description.icon_off
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn the entity on."""
+        await self._write_setting(1)
+        self._attr_is_on = True
```

Comment:
I think it is better to read back the value from the inverter instead of just assuming it has been set correctly.
Use: `await inverter.read_setting(description.key)` or even better request it through the update function I think we should implement or the update_coordinator.

Reply:
will do